proposals/jsr-305-custom-nullability-qualifiers.md

```diff
@@ -0,0 +1,321 @@
+it means that they can be used both as nullable and not-nullable. 
+
+In addition, Java types declarations may be annotated with some of the known 
+annotations to make them have [more precise types](https://github.com/JetBrains/kotlin/blob/master/spec-docs/flexible-java-types.md#more-precise-type-information-from-annotations)
+when used from Kotlin. Let's call this kind of annotating as types enhancement.
+
+Current proposal is supposed to introduce additional ways of types enhancement
+and instruments to control their migration status.
+
+### Type qualifier nicknames
+[`@TypeQualifierNickname`](https://aalmiray.github.io/jsr-305/apidocs/javax/annotation/meta/TypeQualifierNickname.html) 
+annotation from the JSR-305 among others allows introducing new nullability
+annotations, and this proposal suggests to interpret them in Kotlin in the 
+same way as other (built-in) nullability annotations.
+
+The rules are following: an annotation class must be annotated both with 
+`@javax.annotation.meta.TypeQualifierNickname` and `@javax.annotation.NonNull` 
+annotation with an optional argument. The latter annotation may be replaced
+with another type qualifier nickname to `NonNull`, including for example
+`@CheckForNull`.
+
+This new annotation class when being applied to a type container 
+declaration (a value parameter, field or method) must enhance related type
+in the same way it would be done if it was annotated as `@NonNull` with
+corresponding `when` argument. Namely:
+- `When.ALWAYS` makes the type not-nullable
+- `When.MAYBE` makes the type nullable
+- `When.NEVER/When.UNKNOWN` doesn't affect type's nullability
+
+```java
+@TypeQualifierNickname
+@Nonnull(when = When.ALWAYS)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface MyNonnull {
+}
+
+@TypeQualifierNickname
+@CheckForNull // a nickname to another type qualifier nickname
+@Retention(RetentionPolicy.RUNTIME)
+public @interface MyNullable {
+}
+
+interface A {
+    @MyNullable
+    String foo(@MyNonnull String x);
+}
+```
+
+`A::foo` in the example above should be available in Kotlin as `fun foo(x: 
+String): String?`
+
+
+### Type qualifier default
+[`@TypeQualifierDefault`](https://aalmiray.github.io/jsr-305/apidocs/javax/annotation/meta/TypeQualifierDefault.html)
+allows introducing annotations that when being applied define the default
+nullability within the scope of the annotated element.
+
+To make it work new annotation class must be annotated with 
+`@TypeQualifierDefault` and `@NonNull` or any type qualifier nickname to the 
+latter.
+The single argument for the `@TypeQualifierDefault` defines a set of element 
+types for which the annotation may enhance a type. 
+
+When determining immediate nullability of a type the Kotlin compiler should look
+for nullability annotation on the type itself at first.
+- If there is one, then it should use it
+- Otherwise, nullability is determined by the closest annotation marked as
```

Comment:
It may be unclear to users what "closest" means here

Reply:
@udalov How do you think, would the phrase "the innermost enclosing element annotated with `TypeQualifierDefault`" be more clear in the context?

Reply:
Yes, that's more clear